Issue with MatTable using MatSort

I have a problem with ngMocks, MatTableModule and MatSortModule.

If I have the mat-table sort initialization in the `ngOnInit` (or `ngAfterViewInit`, like Angular Material documents), I've got the following error:
```
  TypeError: You provided 'undefined' where a stream was expected. You can provide an Observable, Promise, Array, or Iterable.
```

Component's code:
```typescript
  ...
  datasource = new MatTableDataSource<object>([]);
  @ViewChild(MatSort, { static: true }) mySort: MatSort = new MatSort();

  ngOnInit(): void {
    this.datasource.sort = this.mySort;
  }
  ...
```

If I move the sort initialization in the constructor, it works.
I can use my init in the constructor but on my point of view, it would also work if initialization is present in ngOnInit, right ?

I got the build error
```
Error: src/test.ts:33:1 - error TS2769: No overload matches this call.
  Overload 1 of 3, '(token: InjectionToken<{ streamPropWeNeedToStub: Observable<never>; }>, handler?: (value: { streamPropWeNeedToStub: Observable<never>; }, injector: Injector) => Partial<...>): void', gave the following error.
    Argument of type 'typeof MatSort' is not assignable to parameter of type 'InjectionToken<{ streamPropWeNeedToStub: Observable<never>; }>'.
      Type 'typeof MatSort' is missing the following properties from type 'InjectionToken<{ streamPropWeNeedToStub: Observable<never>; }>': _desc, ɵprov
  Overload 2 of 3, '(token: string, handler?: (value: { streamPropWeNeedToStub: Observable<never>; }, injector: Injector) => Partial<{ streamPropWeNeedToStub: Observable<never>; }>): void', gave the following error.
    Argument of type 'typeof MatSort' is not assignable to parameter of type 'string'.
  Overload 3 of 3, '(def: AnyType<MatSort>, handler?: (value: MatSort, injector: Injector) => void | Partial<MatSort>): void', gave the following error.
    Type '{ streamPropWeNeedToStub: Observable<never>; }' is not assignable to type 'void | Partial<MatSort>'.
      Type '{ streamPropWeNeedToStub: Observable<never>; }' has no properties in common with type 'Partial<MatSort>'.
      ```

--- a/examples/MockBuilder/test.ng-mocks.spec.ts
+++ b/examples/MockBuilder/test.ng-mocks.spec.ts
@@ -128,7 +128,7 @@ describe('MockBuilder:ngMocks', () => {
       expect(serviceMockInstance.getName).toBeDefined();
       expect(serviceMockInstance.getName()).toBeUndefined();
       expect(mocks.has(TOKEN_MOCK)).toBeDefined();
-      expect(mocks.get(TOKEN_MOCK)).toBeUndefined();
+      expect(mocks.get(TOKEN_MOCK)).toBeDefined();
 
       // customize
       const serviceCustomize = mocks.get(ServiceCustomize);

--- a/libs/ng-mocks/src/lib/mock-builder/promise/add-missed-mock-declarations-and-modules.ts
+++ b/libs/ng-mocks/src/lib/mock-builder/promise/add-missed-mock-declarations-and-modules.ts
@@ -21,7 +21,7 @@ export default (ngModule: NgMeta, { mockDef, configDef }: BuilderData): void => 
     }
 
     const mock = ngMocksUniverse.builtProviders.get(def);
-    ngModule.providers.push(mock || { provide: def, useValue: undefined });
+    ngModule.providers.push(mock);
     ngMocksUniverse.touches.add(def);
   }
 };

--- a/libs/ng-mocks/src/lib/mock-builder/promise/try-mock-provider.ts
+++ b/libs/ng-mocks/src/lib/mock-builder/promise/try-mock-provider.ts
@@ -24,7 +24,7 @@ export default (def: any, defValue: Map<any, any>): void => {
       helperUseFactory(def, undefined, existing => createInstance(existing, instance, config, isPipeFunc)),
     );
   } else if (isNgDef(def, 'i')) {
-    ngMocksUniverse.builtProviders.set(def, mockProvider(def));
+    ngMocksUniverse.builtProviders.set(def, mockProvider(def, true));
   }
 
   if (!isNgDef(def) && defValue.has(def)) {
@@ -34,6 +34,6 @@ export default (def: any, defValue: Map<any, any>): void => {
       helperUseFactory(def, undefined, () => instance),
     );
   } else if (!isNgDef(def)) {
-    ngMocksUniverse.builtProviders.set(def, mockProvider(def));
+    ngMocksUniverse.builtProviders.set(def, mockProvider(def, true));
   }
 };

--- a/libs/ng-mocks/src/lib/mock-helper/mock-helper.default-mock.ts
+++ b/libs/ng-mocks/src/lib/mock-helper/mock-helper.default-mock.ts
@@ -1,18 +1,21 @@
 import { InjectionToken, Injector } from '@angular/core';
 
+import { flatten } from '../common/core.helpers';
 import { AnyType } from '../common/core.types';
 import ngMocksUniverse from '../common/ng-mocks-universe';
 
 export default <T>(
-  def: AnyType<T> | InjectionToken<T> | string,
+  def: AnyType<T> | InjectionToken<T> | string | Array<AnyType<T> | InjectionToken<T> | string>,
   callback?: (instance: undefined | T, injector: Injector) => void | Partial<T>,
 ): void => {
   const map = ngMocksUniverse.getOverrides();
-  if (callback) {
-    const set: Set<any> = map.has(def) ? map.get(def) : new Set();
-    set.add(callback);
-    map.set(def, set);
-  } else {
-    map.delete(def);
+  for (const item of flatten(def)) {
+    if (callback) {
+      const set: Set<any> = map.has(item) ? map.get(item) : new Set();
+      set.add(callback);
+      map.set(item, set);
+    } else {
+      map.delete(item);
+    }
   }
 };

--- a/libs/ng-mocks/src/lib/mock-helper/mock-helper.ts
+++ b/libs/ng-mocks/src/lib/mock-helper/mock-helper.ts
@@ -71,6 +71,14 @@ export const ngMocks: {
   defaultMock<T>(def: AnyType<T>, handler?: (value: T, injector: Injector) => void | Partial<T>): void;
 
   /**
+   * @see https://ng-mocks.sudo.eu/api/ngMocks/defaultMock
+   */
+  defaultMock<T = any>(
+    defs: Array<AnyType<any> | InjectionToken<any>>,
+    handler?: (value: undefined | T, injector: Injector) => undefined | Partial<T>,
+  ): void;
+
+  /**
    * @see https://ng-mocks.sudo.eu/api/ngMocks/event
    */
   event(

--- a/libs/ng-mocks/src/lib/mock-module/mock-module.spec.ts
+++ b/libs/ng-mocks/src/lib/mock-module/mock-module.spec.ts
@@ -258,8 +258,22 @@ describe('mockProvider', () => {
   });
 
   it('should return undefined on any token', () => {
-    expect(mockProvider(CUSTOM_TOKEN)).toBeUndefined();
-    expect(mockProvider(HTTP_INTERCEPTORS)).toBeUndefined();
+    const p1: any = mockProvider(CUSTOM_TOKEN, true);
+    expect(p1).toEqual({
+      deps: [Injector],
+      provide: CUSTOM_TOKEN,
+      useFactory: jasmine.anything(),
+    });
+    expect(p1.useFactory()).toEqual(undefined);
+
+    const p2: any = mockProvider(HTTP_INTERCEPTORS, true);
+    expect(p2).toEqual({
+      deps: [Injector],
+      provide: HTTP_INTERCEPTORS,
+      useFactory: jasmine.anything(),
+    });
+    expect(p2.useFactory()).toEqual(undefined);
+
     expect(mockProvider(APP_INITIALIZER)).toBeUndefined();
   });
 

--- a/libs/ng-mocks/src/lib/mock-render/mock-render-factory.ts
+++ b/libs/ng-mocks/src/lib/mock-render/mock-render-factory.ts
@@ -2,7 +2,7 @@ import { DebugElement, Directive, InjectionToken } from '@angular/core';
 import { getTestBed, TestBed } from '@angular/core/testing';
 
 import coreDefineProperty from '../common/core.define-property';
-import { Type } from '../common/core.types';
+import { AnyType, Type } from '../common/core.types';
 import funcImportExists from '../common/func.import-exists';
 import { isNgDef } from '../common/func.is-ng-def';
 import ngMocksUniverse from '../common/ng-mocks-universe';
@@ -16,7 +16,7 @@ import funcReflectTemplate from './func.reflect-template';
 import { DefaultRenderComponent, IMockRenderOptions, MockedComponentFixture } from './types';
 
 interface MockRenderFactory<C = any, F = DefaultRenderComponent<C>> {
-  declaration: Type<never>;
+  declaration: AnyType<never>;
   params: F;
   (): MockedComponentFixture<C, F>;
 }
@@ -123,7 +123,7 @@ function MockRenderFactory<MComponent>(
  * @see https://ng-mocks.sudo.eu/api/MockRender#factory
  */
 function MockRenderFactory<MComponent>(
-  template: Type<MComponent>,
+  template: AnyType<MComponent>,
   params: undefined | null,
   detectChangesOrOptions?: boolean | IMockRenderOptions,
 ): MockRenderFactory<MComponent, MComponent>;
@@ -132,7 +132,7 @@ function MockRenderFactory<MComponent>(
  * @see https://ng-mocks.sudo.eu/api/MockRender#factory
  */
 function MockRenderFactory<MComponent, TComponent extends object>(
-  template: Type<MComponent>,
+  template: AnyType<MComponent>,
   params: TComponent,
   detectChangesOrOptions?: boolean | IMockRenderOptions,
 ): MockRenderFactory<MComponent, TComponent>;
@@ -141,7 +141,7 @@ function MockRenderFactory<MComponent, TComponent extends object>(
  * @see https://ng-mocks.sudo.eu/api/MockRender#factory
  */
 function MockRenderFactory<MComponent, TComponent extends object = Record<keyof any, any>>(
-  template: Type<MComponent>,
+  template: AnyType<MComponent>,
   params: TComponent,
   detectChangesOrOptions?: boolean | IMockRenderOptions,
 ): MockRenderFactory<MComponent, TComponent>;
@@ -151,7 +151,7 @@ function MockRenderFactory<MComponent, TComponent extends object = Record<keyof 
  *
  * @see https://ng-mocks.sudo.eu/api/MockRender#factory
  */
-function MockRenderFactory<MComponent>(template: Type<MComponent>): MockRenderFactory<MComponent, MComponent>;
+function MockRenderFactory<MComponent>(template: AnyType<MComponent>): MockRenderFactory<MComponent, MComponent>;
 
 /**
  * An empty string does not have point.
@@ -195,7 +195,7 @@ function MockRenderFactory<MComponent, TComponent extends Record<keyof any, any>
 ): MockRenderFactory<MComponent, TComponent>;
 
 function MockRenderFactory<MComponent, TComponent extends Record<keyof any, any>>(
-  template: string | Type<MComponent> | InjectionToken<MComponent>,
+  template: string | AnyType<MComponent> | InjectionToken<MComponent>,
   params?: TComponent,
   flags: boolean | IMockRenderOptions = true,
 ): any {

--- a/libs/ng-mocks/src/lib/mock-render/mock-render.ts
+++ b/libs/ng-mocks/src/lib/mock-render/mock-render.ts
@@ -1,7 +1,7 @@
 import { InjectionToken } from '@angular/core';
 import { ComponentFixture } from '@angular/core/testing';
 
-import { Type } from '../common/core.types';
+import { AnyType } from '../common/core.types';
 
 import { MockRenderFactory } from './mock-render-factory';
 import { IMockRenderOptions, MockedComponentFixture } from './types';
@@ -19,7 +19,7 @@ function MockRender<MComponent>(
  * @see https://ng-mocks.sudo.eu/api/MockRender
  */
 function MockRender<MComponent>(
-  template: Type<MComponent>,
+  template: AnyType<MComponent>,
   params: undefined | null,
   detectChangesOrOptions?: boolean | IMockRenderOptions,
 ): MockedComponentFixture<MComponent, MComponent>;
@@ -28,7 +28,7 @@ function MockRender<MComponent>(
  * @see https://ng-mocks.sudo.eu/api/MockRender
  */
 function MockRender<MComponent, TComponent extends object>(
-  template: Type<MComponent>,
+  template: AnyType<MComponent>,
   params: TComponent,
   detectChangesOrOptions?: boolean | IMockRenderOptions,
 ): MockedComponentFixture<MComponent, TComponent>;
@@ -37,7 +37,7 @@ function MockRender<MComponent, TComponent extends object>(
  * @see https://ng-mocks.sudo.eu/api/MockRender
  */
 function MockRender<MComponent, TComponent extends object = Record<keyof any, any>>(
-  template: Type<MComponent>,
+  template: AnyType<MComponent>,
   params: TComponent,
   detectChangesOrOptions?: boolean | IMockRenderOptions,
 ): MockedComponentFixture<MComponent, TComponent>;
@@ -47,7 +47,7 @@ function MockRender<MComponent, TComponent extends object = Record<keyof any, an
  *
  * @see https://ng-mocks.sudo.eu/api/MockRender
  */
-function MockRender<MComponent>(template: Type<MComponent>): MockedComponentFixture<MComponent, MComponent>;
+function MockRender<MComponent>(template: AnyType<MComponent>): MockedComponentFixture<MComponent, MComponent>;
 
 /**
  * An empty string does not have point.
@@ -91,7 +91,7 @@ function MockRender<MComponent, TComponent extends Record<keyof any, any> = Reco
 ): MockedComponentFixture<MComponent, TComponent>;
 
 function MockRender<MComponent, TComponent extends Record<keyof any, any>>(
-  template: string | Type<MComponent> | InjectionToken<MComponent>,
+  template: string | AnyType<MComponent> | InjectionToken<MComponent>,
   params?: TComponent,
   flags: boolean | IMockRenderOptions = true,
 ): any {

--- a/libs/ng-mocks/src/lib/mock-service/helper.resolve-provider.ts
+++ b/libs/ng-mocks/src/lib/mock-service/helper.resolve-provider.ts
@@ -102,12 +102,6 @@ const createPredefinedMockProvider = (provider: any, provide: any): any => {
     if (mockDef === provide) {
       return provider;
     }
-    if (mockDef === undefined) {
-      return {
-        provide,
-        useValue: undefined,
-      };
-    }
 
     return mockDef;
   }

--- a/libs/ng-mocks/src/lib/mock-service/mock-provider.ts
+++ b/libs/ng-mocks/src/lib/mock-service/mock-provider.ts
@@ -84,9 +84,9 @@ const createMockProvider = (provider: any, provide: any, cacheProviders?: Map<an
 // The main problem is that providing undefined to HTTP_INTERCEPTORS and others breaks their code.
 // If a testing module / component requires omitted tokens then they should be provided manually
 // during creation of TestBed module.
-const handleProvider = (provider: any, provide: any) => {
+const handleProvider = (provider: any, provide: any, useFactory: boolean) => {
   if (provide === provider) {
-    return undefined;
+    return useFactory ? helperUseFactory(provider, () => undefined) : undefined;
   }
   if (provider.multi) {
     ngMocksUniverse.config.get('ngMocksMulti')?.add(provide);
@@ -115,7 +115,7 @@ const isNeverMockFunction = (provide: any): boolean =>
 const isNeverMockToken = (provide: any): boolean =>
   isNgInjectionToken(provide) && neverMockToken.indexOf(provide.toString()) !== -1;
 
-export default function (provider: any): Provider | undefined {
+export default function (provider: any, useFactory = false): Provider | undefined {
   const provide = funcGetProvider(provider);
 
   if (isNeverMockFunction(provide)) {
@@ -134,5 +134,5 @@ export default function (provider: any): Provider | undefined {
     return cacheProviders.get(provide);
   }
 
-  return createMockProvider(provider, provide, cacheProviders) || handleProvider(provider, provide);
+  return createMockProvider(provider, provide, cacheProviders) || handleProvider(provider, provide, useFactory);
 }

--- a/tests/issue-568/test.spec.ts
+++ b/tests/issue-568/test.spec.ts
@@ -1,0 +1,82 @@
+import { Injectable, InjectionToken } from '@angular/core';
+import { TestBed } from '@angular/core/testing';
+import {
+  MockBuilder,
+  MockProvider,
+  MockRender,
+  ngMocks,
+} from 'ng-mocks';
+
+@Injectable()
+class Service1 {
+  public readonly name = 'service1';
+}
+
+@Injectable()
+abstract class Service2 {
+  public readonly name = 'service2';
+}
+
+const TOKEN = new InjectionToken<{ name: string }>('TOKEN');
+
+ngMocks.defaultMock([Service1, Service2, TOKEN], () => ({
+  name: 'mock',
+}));
+
+describe('issue-568', () => {
+  describe('MockBuilder', () => {
+    ngMocks.faster();
+
+    beforeAll(() =>
+      MockBuilder().mock(Service1).mock(Service2).mock(TOKEN),
+    );
+
+    it('mocks service1', () => {
+      expect(
+        MockRender(Service1).point.componentInstance.name,
+      ).toEqual('mock');
+    });
+
+    it('mocks service2', () => {
+      expect(
+        MockRender(Service2).point.componentInstance.name,
+      ).toEqual('mock');
+    });
+
+    it('mocks token', () => {
+      expect(MockRender(TOKEN).point.componentInstance.name).toEqual(
+        'mock',
+      );
+    });
+  });
+
+  describe('TestBed', () => {
+    beforeEach(() =>
+      TestBed.configureTestingModule({
+        providers: [
+          MockProvider(Service1),
+          MockProvider(Service2),
+          MockProvider(TOKEN),
+        ],
+      }).compileComponents(),
+    );
+
+    it('mocks service1', () => {
+      expect(
+        MockRender(Service1).point.componentInstance.name,
+      ).toEqual('mock');
+    });
+
+    it('mocks service2', () => {
+      expect(
+        MockRender(Service2).point.componentInstance.name,
+      ).toEqual('mock');
+    });
+
+    it('mocks token', () => {
+      expect(MockRender(TOKEN).point.componentInstance.name).toEqual(
+        'mock',
+      );
+    });
+  });
+});

